worldedit-core/src/main/java/com/sk89q/worldedit/extent/reorder/ChunkBatchingExtent.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * WorldEdit, a Minecraft world manipulation toolkit
+ * Copyright (C) sk89q <http://www.sk89q.com>
+ * Copyright (C) WorldEdit team and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.extent.reorder;
+
+import com.sk89q.worldedit.BlockVector2D;
+import com.sk89q.worldedit.Vector;
+import com.sk89q.worldedit.Vector2D;
+import com.sk89q.worldedit.WorldEditException;
+import com.sk89q.worldedit.extent.AbstractDelegateExtent;
+import com.sk89q.worldedit.extent.Extent;
+import com.sk89q.worldedit.function.operation.Operation;
+import com.sk89q.worldedit.function.operation.RunContext;
+import com.sk89q.worldedit.function.operation.SetLocatedBlocks;
+import com.sk89q.worldedit.util.collection.LocatedBlockList;
+import com.sk89q.worldedit.world.block.BlockStateHolder;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * A special extent that batches changes into Minecraft chunks. This helps
+ * improve the speed of setting the blocks, since chunks do not need to be
+ * loaded repeatedly, however it does take more memory due to caching the
+ * blocks.
+ */
+public class ChunkBatchingExtent extends AbstractDelegateExtent {
+
+    /**
+     * Comparator optimized for sorting chunks by the region file they reside
+     * in. This allows for file caches to be used while loading the chunk.
+     */
+    private static final Comparator<Vector2D> REGION_OPTIMIZED_SORT =
+            Comparator.<Vector2D, Vector2D>comparing(vec -> vec.divide(32).floor(), Vector2D.COMPARING_GRID_ARRANGEMENT)
+                    .thenComparing(Vector2D.COMPARING_GRID_ARRANGEMENT);
```

Comment:
Smart, I tried doing it just based off the chunk, didn't seem to be as beneficial. Granted that was also in Sponge and I think the overhead of cause tracking was obliterating any optimizations :cry:.

Reply:
Yea, I'm not sure how much it speeds us up to sort by region, but the overall change was from `50k` to `70k` on my machine, or about 40% faster overall. Obviously this may vary, but I think it should be effective overall.

@ryantheleach I missed your comment but here's the answer :sweat_smile:

Reply:
Thanks.